### ISSUE TITLE
CatBoost official benchmark examples use test data as the 'validation set'? 

### ISSUE THREAD

**Issue by SSLPP** [catboost/catboost#1422]
Problem: Is CatBoost Official Benchmark examples using test data as 'eval_set'? 
catboost version:
Operating System:
CPU:
GPU:

While I was going through [CatBoost Official Benchmark examples](https://github.com/catboost/benchmarks/tree/master/quality_benchmarks) I noticed that the code is using the test data as the 'eval_set' for the final validation. Is this the correct approach? I am under the impression that 'test data' is reserved for final testing. 

Code from [here](https://github.com/catboost/benchmarks/blob/master/quality_benchmarks/notebooks/comparison_amazon.ipynb). Focus on **ITEM 1**

```python
def run_experiment(Experiment, title):
    experiment = Experiment(learning_task, max_hyperopt_evals=max_hyperopt_evals,
                            n_estimators=n_estimators)
    cv_pairs, (dtrain, dtest) = experiment.split_and_preprocess(X_train.copy(), y_train, 
                                                                X_test.copy(), y_test, 
                                                                cat_cols, n_splits=5)
 
    default_cv_result = experiment.run_cv(cv_pairs)
    experiment.print_result(default_cv_result, 'Default {} result on cv'.format(title))

    print('\n~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n')

    default_test_losses = []
    for seed in range(5):
        test_result = experiment.run_test(dtrain, dtest, X_test, params=default_cv_result['params'],
                                          n_estimators=default_cv_result['best_n_estimators'], seed=seed)
        default_test_losses.append(test_result['loss'])
        print 'For seed=%d Test\'s %s : %.5f' % (seed, experiment.metric, default_test_losses[-1])
    print '\nTest\'s %s mean: %.5f, Test\'s %s std: %.5f' % (experiment.metric, np.mean(default_test_losses),
                                                             experiment.metric, np.std(default_test_losses))

    print('\n~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n')
    print('Hyperopt iterations:\n\n')

    tuned_cv_result = experiment.optimize_params(cv_pairs)

    print('\n~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n')

    experiment.print_result(tuned_cv_result, 'Tuned {} result on cv'.format(title))

    print('\n~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n')

    tuned_test_losses = []
    for seed in range(5):
        ####################################################
        # ITEM 1
        ########################################################
        test_result = experiment.run_test(dtrain, dtest, X_test, params=tuned_cv_result['params'],
                                          n_estimators=tuned_cv_result['best_n_estimators'], seed=seed)
        tuned_test_losses.append(test_result['loss'])
        print 'For seed=%d Test\'s %s : %.5f' % (seed, experiment.metric, tuned_test_losses[-1])
    print '\nTest\'s %s mean: %.5f, Test\'s %s std: %.5f' % (experiment.metric, np.mean(tuned_test_losses),
                                                             experiment.metric, np.std(tuned_test_losses))

    return np.mean(default_test_losses), np.mean(tuned_test_losses)
```

The `run_test` function in 'experiment.py': Focus on Item #2

```python
    def run_test(self, dtrain, dtest, X_test=None, params=None, n_estimators=None, custom_metric=None, seed=0):
        params = params or self.best_params or self.default_params
        n_estimators = n_estimators or self.best_n_estimators or self.n_estimators
        params = self.preprocess_params(params)
        start_time = time.time()
        ####################################################
        # ITEM 2
        ########################################################
        bst, evals_result = self.fit(params, dtrain, dtest, n_estimators, seed=seed)
        eval_time = time.time() - start_time
        preds = self.predict(bst, dtest, X_test)

        result = {'loss': evals_result[-1], 'bst': bst, 'n_estimators': n_estimators,
                  'eval_time': eval_time, 'status': STATUS_OK,  'params': params.copy(),
                  'preds': preds}

        if custom_metric is not None:
            if type(custom_metric) is not dict:
                raise TypeError("custom_metric argument should be dict")
            pred = self.predict(bst, dtest, X_test)
            for title, func in custom_metric.iteritems():
                score = func(dtest.get_label(), pred, sample_weight=None) # TODO weights
                result[title] = score

        return result
```

The fit function from `catboost_experiment.py` in 'catboost_experiment.py' . Focus on Item 3

```python
    def fit(self, params, dtrain, dtest, n_estimators, seed=0):
        params.update({"iterations": n_estimators})
        params.update({"random_seed": seed})
        bst = CatBoost(params)
        ####################################################
        # ITEM 3
        ########################################################
        bst.fit(dtrain, eval_set=dtest)
        with open("test_error.tsv", "r") as f:
            results = np.array(map(lambda x: float(x.strip().split()[-1]), f.readlines()[1:]))
        
        return bst, results
```

The code uses `dtest` set as the `validation` set. Is this the correct approach?

CatBoost `fit` uses `eval_set` as the validation dataset for 'overfitting detector', 'best iteration selection' , and "monitoring metrics' changes." 

I'm confused, please correct me if I'm missing something.
